### ISSUE TITLE
`StartWith[...]Lifetime` should dispose the `Application` instance

### ISSUE THREAD

**Issue by Tyrrrz** [AvaloniaUI/Avalonia#15142]
### Is your feature request related to a problem? Please describe.

My implementation of `Application` has a `Dispose` method that disposes the inner `ServiceProvider` that contains various domain-specific services and view models.

`Application` is not disposed automatically by Avalonia. So I have to run the app like this:

```csharp
// Program.cs

public static AppBuilder BuildAvaloniaApp() =>
    AppBuilder.Configure<App>().UsePlatformDetect().LogToTrace();

[STAThread]
public static int Main(string[] args)
{
    var builder = BuildAvaloniaApp();

    try
    {
        return builder.StartWithClassicDesktopLifetime(args);
    }
    finally
    {
        // Clean up after application shutdown
        if (builder.Instance is IDisposable disposableApp)
            disposableApp.Dispose();
    }
}
```

This is a bit clunky. It would be much nicer if the underlying plumbing code automatically checked if the produced `Application` instance implements `IDisposable` and executes it if so. The `AppBuilder` is responsible for instantiating the object, so it makes sense it would also be responsible for its cleanup.

### Describe the solution you'd like

I want the following code to execute `Dispose()` on my application automatically:

```csharp
// Program.cs

public static AppBuilder BuildAvaloniaApp() =>
    AppBuilder.Configure<App>().UsePlatformDetect().LogToTrace();

[STAThread]
public static int Main(string[] args) =>
    BuildAvaloniaApp().StartWithClassicDesktopLifetime(args);
```

### Describe alternatives you've considered

The current alternative I use is described above. I also looked into implementing a custom Lifetime, but it looks like it's not very easy.

**Comment by maxkatz6** [comment]
I don't think we can reliably have that implemented. Sure, for ClassicDesktopLifetime it would be straightforward. But not all lifeimes have even a concept of being closed. And embedding scenarios (i.e. - without lifetime) won't have that either. 

Promising that Application is going to be disposed on some platforms, will create similar expectations about other platforms.

**Comment by Tyrrrz** [comment]
Could that be an option that is specifically configured for classic desktop lifetime then? Something like:

```csharp
[STAThread]
public static int Main(string[] args) =>
    BuildAvaloniaApp().StartWithClassicDesktopLifetime(args, disposeApplication: true);
```

**Comment by wsficke** [comment]
@Tyrrrz here's the recipe.  You should use a custom app lifetime to hook your DI container.

[// Program.cs:

            BuildAvaloniaApp()
                .Start(App.AppMain, args);](url)
                
// App.axaml.cs
```
        public static void AppMain(Application app, string[] commandLineArguments)
        {
            // build your DI container
            [proprietary code omitted]
            
            app.Run(ShutdownRequestCts.Token); // Run, observing a CancellationTokenSource that breaks the event loop
            
            
            // dispose your DI container
            [proprietary code omitted]
        }
        
 ```

**Comment by jp2masa** [comment]
> The previewer depends on having a 0-parameter constructor on all your ViewModels. So if you want to use the previewer with real data, constructor injection is out.

You can use constructor injection if you implement an approach as the one described in https://github.com/AvaloniaUI/Avalonia/discussions/13743#discussioncomment-7686499, for example.
